src/proto/connection.rs

```diff
@@ -397,3 +396,13 @@ where
         self.ping_pong.ping_shutdown();
     }
 }
+
+impl<T, P, B> Drop for Connection<T, P, B>
+where
+    P: Peer,
+    B: IntoBuf,
+{
+    fn drop(&mut self) {
+        self.streams.recv_eof();
```

Comment:
If this is dropped because the thread panicked, this will trigger an abort (`recv_eof` unwraps a lock).

Reply:
Ok, we can have `recv_eof()` return `Err` instead of unwrapping the poisoned mutex.